pkg/agent/endpoints/workload/handler_test.go

```diff
@@ -321,6 +321,122 @@ func TestFetchX509Bundles(t *testing.T) {
 	}
 }
 
+func TestFetchX509Bundles_MultipleUpdates(t *testing.T) {
+	ca := testca.New(t, td)
+	x509SVID := ca.CreateX509SVID(workloadID)
+
+	bundle := ca.Bundle()
+	bundleX509 := x509util.DERFromCertificates(bundle.X509Authorities())
+
+	otherBundle := testca.New(t, td).Bundle()
+	otherBundleX509 := x509util.DERFromCertificates(otherBundle.X509Authorities())
+
+	updates := []*cache.WorkloadUpdate{
+		{
+			Identities: []cache.Identity{
+				identityFromX509SVID(x509SVID),
+			},
+			Bundle: utilBundleFromBundle(t, bundle),
+		},
+		{
+			Identities: []cache.Identity{
+				identityFromX509SVID(x509SVID),
+			},
+			Bundle: utilBundleFromBundle(t, otherBundle),
+		},
+	}
+
+	expectResp := []*workloadPB.X509BundlesResponse{
+		{
+			Bundles: map[string][]byte{
+				bundle.TrustDomain().IDString(): bundleX509,
+			},
+		},
+		{
+			Bundles: map[string][]byte{
+				bundle.TrustDomain().IDString(): otherBundleX509,
+			},
+		},
+	}
+
+	params := testParams{
+		CA:                            ca,
+		Updates:                       updates,
+		AttestErr:                     nil,
+		ExpectLogs:                    nil,
+		AllowUnauthenticatedVerifiers: false,
+	}
+
+	runTest(t, params,
+		func(ctx context.Context, client workloadPB.SpiffeWorkloadAPIClient) {
+			stream, err := client.FetchX509Bundles(ctx, &workloadPB.X509BundlesRequest{})
+			require.NoError(t, err)
+
+			resp, err := stream.Recv()
+			spiretest.RequireGRPCStatus(t, err, codes.OK, "")
+			spiretest.RequireProtoEqual(t, expectResp[0], resp)
+
+			resp, err = stream.Recv()
+			spiretest.RequireGRPCStatus(t, err, codes.OK, "")
+			spiretest.RequireProtoEqual(t, expectResp[1], resp)
+		})
+}
+
+func TestFetchX509Bundles_SpuriousUpdates(t *testing.T) {
+	ca := testca.New(t, td)
+	x509SVID := ca.CreateX509SVID(workloadID)
+
+	bundle := ca.Bundle()
+	bundleX509 := x509util.DERFromCertificates(bundle.X509Authorities())
+
+	updates := []*cache.WorkloadUpdate{
+		{
+			Identities: []cache.Identity{
+				identityFromX509SVID(x509SVID),
+			},
+			Bundle: utilBundleFromBundle(t, bundle),
+		},
+		{
+			Identities: []cache.Identity{
+				identityFromX509SVID(x509SVID),
+			},
+			Bundle: utilBundleFromBundle(t, bundle),
+		},
+	}
+
+	expectResp := []*workloadPB.X509BundlesResponse{
+		{
+			Bundles: map[string][]byte{
+				bundle.TrustDomain().IDString(): bundleX509,
+			},
+		},
+	}
+
+	params := testParams{
+		CA:                            ca,
+		Updates:                       updates,
+		AttestErr:                     nil,
+		ExpectLogs:                    nil,
+		AllowUnauthenticatedVerifiers: false,
+	}
+
+	runTest(t, params,
+		func(ctx context.Context, client workloadPB.SpiffeWorkloadAPIClient) {
+			timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
```

Comment:
Hmm, this will add 5 seconds to the unit test, correct? I wonder if there is another way we can test this deterministically without introducing this time. I have two reasons:
1. This kind of change seems small but adds up in the aggregate, increasing our already terrible CI/CD pipeline times
2. When running in CI/CD, due to the often resource constrained running environment, waiting for small time values like this doesn't always mean that the code we're waiting on had the chance to do everything. For example, in this case, it could be that another response was going to be sent down the stream, but the code was executing slowly enough that it would have taken longer than five seconds to do so. In which case we'd have a false positive indication of success.

One way we might be able to test this deterministically is by creating a test hook in the stream response loop so we can determine that we've received an update from the cache, we've sent the response (or not if there was a change) and are waiting for the next update. I think we could probably implement that hook without impacting the production code by adding some code to the fakeSubscriber, which gets a call to `Update` in the stream update loop.

Reply:
I'm looking into how to test it with a hook in the stream update loop.

As an alternative, what do you think about adding a debug log?
```
if proto.Equal(resp, previousResponse) {
    log.Debug("spurious update")
    return previousResponse, nil
}
```
and add it to expectedLogs in the test.